File: src/bmsx/utils/scratchbatch.ts
// Lightweight, reusable scratch collections for per-frame batching.
//
// Goals:
// - Avoid per-frame allocations by retaining backing storage.
// - Offer a simple, consistent API across systems.
// - Optional pooling for item reuse via Pool<T>.
// - Excluded from savegame serialization.

import { excludeclassfromsavegame } from 'bmsx/serializer/serializationhooks';
import type { Pool } from './pool';

@excludeclassfromsavegame
export class ScratchBatch<T> implements Iterable<T> {
    private items: T[] = [];
    private _size = 0;

    constructor(initialCapacity = 0) {
        if (initialCapacity > 0) this.items.length = initialCapacity;
    }

    get size(): number { return this._size; }

    // For convenience when migrating legacy arrays (read-only length)
    get length(): number { return this._size; }

    clear(): void { this._size = 0; }

    push(v: T): void { this.items[this._size++] = v; }

    get(index: number): T { return this.items[index]; }

    forEach(cb: (item: T, index: number) => void): void {
        for (let i = 0; i < this._size; i++) cb(this.items[i], i);
    }

    // Sort only the active window; avoids copying.
    sort(compareFn: (a: T, b: T) => number): void {
        const prevLen = this.items.length;
        this.items.length = this._size;
        this.items.sort(compareFn);
        this.items.length = prevLen;
    }

    [Symbol.iterator](): Iterator<T> {
        let i = 0;
        const n = this._size;
        const arr = this.items;
        return {
            next(): IteratorResult<T> {
                if (i < n) return { value: arr[i++], done: false };
                return { value: undefined as T, done: true };
            },
        };
    }
}

// Optional pooled variant: acquires item slots from a Pool<T> to avoid per-frame object creation.
@excludeclassfromsavegame
export class ScratchBatchPooled<T> implements Iterable<T> {
    private list = new ScratchBatch<T>();

    constructor(private pool: Pool<T>) { }

    get size(): number { return this.list.size; }
    get length(): number { return this.list.length; }

    // Acquire a slot from the pool and include it in the active window.
    next(): T {
        const slot = this.pool.acquire();
        if (!slot) throw new Error('ScratchBatchPooled: pool exhausted');
        this.list.push(slot);
        return slot;
    }

    get(index: number): T { return this.list.get(index); }

    clear(): void {
        // Release all active items back to the pool (idempotent per item as Pool.release is safe).
        for (const item of this.list) this.pool.release(item);
        this.list.clear();
    }

    forEach(cb: (item: T, index: number) => void): void { this.list.forEach(cb); }
    sort(compareFn: (a: T, b: T) => number): void { this.list.sort(compareFn); }
    [Symbol.iterator](): Iterator<T> { return this.list[Symbol.iterator](); }
}

File: src/bmsx/res/bios/util/scratchbatch.lua
-- scratchbatch.lua
-- lightweight reusable scratch collections for per-frame batching

local scratchbatch = {}
scratchbatch.__index = scratchbatch

function scratchbatch.new(initial_capacity, fill_value)
	local items = {}
	local count = initial_capacity or 0
	local i = 0
	while i < count do
		i = i + 1
		items[i] = fill_value
	end
	return setmetatable({
		items = items,
		size = 0,
		length = 0,
	}, scratchbatch)
end

function scratchbatch:clear()
	self.size = 0
	self.length = 0
end

function scratchbatch:push(value)
	local next_index = self.size + 1
	self.items[next_index] = value
	self.size = next_index
	self.length = next_index
end

function scratchbatch:get(index)
	return self.items[index]
end

function scratchbatch:reserve(min_capacity, fill_value)
	local items = self.items
	while #items < min_capacity do
		items[#items + 1] = fill_value
	end
	return items
end

function scratchbatch:for_each(callback)
	for i = 1, self.size do
		callback(self.items[i], i)
	end
end

function scratchbatch:iter()
	local index = 0
	return function()
		index = index + 1
		if index <= self.size then
			return self.items[index], index
		end
	end
end

return scratchbatch
